thanks looks like there's an fd use after close bug:

```
>               warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))
E               pytest.PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1
E               
E               Traceback (most recent call last):
E                 File "/Users/runner/hostedtoolcache/PyPy/3.8.16/x64/lib/pypy3.8/threading.py", line 932, in _bootstrap_inner
E                   self.run()
E                 File "/Users/runner/hostedtoolcache/PyPy/3.8.16/x64/lib/pypy3.8/threading.py", line 870, in run
E                   self._target(*self._args, **self._kwargs)
E                 File "/Users/runner/work/hupper/hupper/.tox/py/lib/pypy3.8/site-packages/hupper/ipc.py", line 167, in _read_loop
E                   packet = self._recv_packet()
E                 File "/Users/runner/work/hupper/hupper/.tox/py/lib/pypy3.8/site-packages/hupper/ipc.py", line 144, in _recv_packet
E                   chunk = os.read(self.r_fd, self._packet_len.size)
E               OSError: [Errno 9] Bad file descriptor
```

--- a/src/hupper/ipc.py
+++ b/src/hupper/ipc.py
@@ -1,3 +1,4 @@
+import errno
 import io
 import os
 import pickle
@@ -135,8 +136,11 @@ class Connection(object):
         self.reader_thread.start()
 
     def close(self):
-        close_fd(self.r_fd)
-        close_fd(self.w_fd)
+        self.r_fd, r_fd = -1, self.r_fd
+        self.w_fd, w_fd = -1, self.r_fd
+
+        close_fd(r_fd)
+        close_fd(w_fd)
         self.on_recv = None
 
     def _recv_packet(self):
@@ -170,6 +174,9 @@ class Connection(object):
                 self._on_recv(packet)
         except EOFError:
             pass
+        except OSError as e:
+            if e.errno != errno.EBADF:
+                raise
         self._on_recv(None)
 
     def _write_packet(self, data):

--- a/src/hupper/ipc.py
+++ b/src/hupper/ipc.py
@@ -137,7 +137,7 @@ class Connection(object):
 
     def close(self):
         self.r_fd, r_fd = -1, self.r_fd
-        self.w_fd, w_fd = -1, self.r_fd
+        self.w_fd, w_fd = -1, self.w_fd
 
         close_fd(r_fd)
         close_fd(w_fd)

--- a/tests/test_ipc.py
+++ b/tests/test_ipc.py
@@ -1,21 +1,36 @@
 import queue
 
-from hupper.ipc import Pipe
+from hupper.ipc import Pipe, spawn
+
+
+def echo(pipe):
+    q = queue.Queue()
+    pipe.activate(q.put)
+    msg = q.get()
+    while msg is not None:
+        pipe.send(msg)
+        msg = q.get()
+    pipe.close()
+    pipe.reader_thread.join()
 
 
 def test_ipc_close():
     c1, c2 = Pipe()
     c1_q = queue.Queue()
-    c2_q = queue.Queue()
     c1.activate(c1_q.put)
-    c2.activate(c2_q.put)
 
-    c1.send("hello")
-    c2.send("world")
-    assert c2_q.get() == "hello"
-    assert c1_q.get() == "world"
+    with spawn(
+        __name__ + '.echo',
+        kwargs={"pipe": c2},
+        pass_fds=[c2.r_fd, c2.w_fd],
+    ) as proc:
+        try:
+            c2.close()
 
-    c1.close()
-    c2.close()
-    c1.reader_thread.join()
-    c2.reader_thread.join()
+            c1.send("hello world")
+            assert c1_q.get() == "hello world"
+
+            c1.close()
+            c1.reader_thread.join()
+        finally:
+            proc.terminate()

